test/PoolManager.t.sol

```diff
@@ -738,6 +794,224 @@ contract PoolManagerTest is Test, Deployers, GasSnapshot {
+            uint128 reservesOfSpecified = uint128(specifiedCurrency.balanceOf(address(manager)));
+            hookDeltaSpecified = int128(bound(hookDeltaSpecified, type(int128).min + 1, int128(reservesOfSpecified)));
+            DeltaReturningHook(hookAddr).setDeltaSpecified(hookDeltaSpecified);
+        }
+
+        // setup swap variables
+        PoolSwapTest.TestSettings memory testSettings =
+            PoolSwapTest.TestSettings({takeClaims: false, settleUsingBurn: false});
+        IPoolManager.SwapParams memory params = IPoolManager.SwapParams({
+            zeroForOne: zeroForOne,
+            amountSpecified: amountSpecified,
+            sqrtPriceLimitX96: (zeroForOne ? MIN_PRICE_LIMIT : MAX_PRICE_LIMIT)
+        });
+
+        // ------------------------ FUZZING CASES ------------------------
+        // with an amount specified of 0: the trade reverts
+        if (amountSpecified == 0) {
+            vm.expectRevert(IPoolManager.SwapAmountCannotBeZero.selector);
+            swapRouter.swap(key, params, testSettings, ZERO_BYTES);
+
+            // trade is exact input of n:, the hook cannot TAKE (+ve hookDeltaSpecified) more than n in input
+            // otherwise the user would have to send more than n in input
+        } else if (isExactIn && (hookDeltaSpecified > -amountSpecified)) {
+            vm.expectRevert(Hooks.HookDeltaExceedsSwapAmount.selector);
+            swapRouter.swap(key, params, testSettings, ZERO_BYTES);
+
+            // exact output of n: the hook cannot GIVE (-ve hookDeltaSpecified) more than n in output
+            // otherwise the user would receive more than n in output
+        } else if (!isExactIn && (amountSpecified < -hookDeltaSpecified)) {
+            vm.expectRevert(Hooks.HookDeltaExceedsSwapAmount.selector);
+            swapRouter.swap(key, params, testSettings, ZERO_BYTES);
+
+            // successful swaps !
+        } else {
+            uint256 balanceThisBefore = specifiedCurrency.balanceOf(address(this));
+            uint256 balanceHookBefore = specifiedCurrency.balanceOf(hookAddr);
+
+            BalanceDelta delta = swapRouter.swap(key, params, testSettings, ZERO_BYTES);
+            int128 deltaSpecified = (zeroForOne == isExactIn) ? delta.amount0() : delta.amount1();
+
+            uint256 balanceThisAfter = specifiedCurrency.balanceOf(address(this));
+            uint256 balanceHookAfter = specifiedCurrency.balanceOf(hookAddr);
+
+            // in all cases the hook gets what they took, and the user gets the swap's output delta (checked more below)
+            assertEq(
+                balanceHookBefore.toInt256() + hookDeltaSpecified,
+                balanceHookAfter.toInt256(),
+                "hook balance change incorrect"
+            );
+            assertEq(
+                balanceThisBefore.toInt256() + deltaSpecified,
+                balanceThisAfter.toInt256(),
+                "swapper balance change incorrect"
+            );
```

Comment:
could we add an assert to the pool manager balance? so the pool manager 0 and 1 balance are the current plus the returned deltas from swap?